### DISCUSSION TITLE
Is there a better project management method than maintaining the todo.txt manually?

### DISCUSSION THREAD

**Discussion by jiangyinzuo** [vim/vim#14974] [General]
Vim's todo.txt(`:h todo`) is too long to maintain. Many Github Issues/PRs have already been closed but have not been removed from the todo.txt.
How should we deal with this legacy todo.txt? Should we adopt better project management methods based on the functionalities of GitHub?

**Comment by yegappan** [comment]
Do you the list of items in todo.txt that can be removed?  I don't think there are too many items in todo.txt that overlap with the Github issues.

**Reply by jiangyinzuo** [reply]
> Do you the list of items in todo.txt that can be removed? I don't think there are too many items in todo.txt that overlap with the Github issues.

I manually checked PR #12428, #12192, #12166 and #12032 in todo.txt, they are all closed. But I have not checked the reason why they are closed. Are they closed as finished, outdated or more work needed?

`grep -c '#[0-9]' runtime/doc/todo.txt` has 283 matched items. `grep -c 'PR' runtime/doc/todo.txt` has 15 matched items.

There are too many items to check one by one.

**Comment by yegappan** [comment]
Bram used to periodically update the todo.txt file.  We haven't done that in a while.  I will try to remove the closed items from todo.txt.

**Comment by jiangyinzuo** [comment]
Yes. I think updating the todo.txt manually is easy to forget, and writing a CI bot for todo.txt seems to be troublesome.

So can we utilize some functionalities of GitHub (milestones, projects, labels, ...) to automatically and visually manage issues and PRs, and leave todo.txt as a roadmap without trivial details?

There are already 2 milestones (backlog and vim-9.1) in vim/vim repository, they are helpful for developers and users to categorize and track issues & PRs.

**Comment by yegappan** [comment]
We already use the Github functionalities to track issues, PRs, label issues and milestones.  The todo.txt file is used more for people who don't know how to use Github to lookup issues and planned features (and in the future if we migrate away from Github for some reason).

**Comment by georgalis** [comment]
the todo is far more organized, and readable holistically than the github issue format, I presume it was developed before github (pre-2010?). For minor features and improvements to work on, it is easier to review for items of interest, IMHO. I'm not suggesting synchronizing, or even populating both simultaneously; and not being an author or maintainer, I'm reluctant to suggest a workflow, but maybe a regex script to check if numbers are "is:closed" and similar efforts to prune it gradually, as github issues are closed? Are there a lot of resolved items in the todo? It would be annoying looking for tasks there and finding the issues actually resolved, but that would be discovered quickly, and an easy PR for the todo. I'm just saying, the todo has value outside the issue tracker, if it needs pruning that effort would be helpful, but an effort to eliminate it would not be. IMHO, pruning the todo per resolved issues would be helpful, perhaps a note to that effect in the first section, as a first step, as well as updating that section about github discussions, etc.

(BTW aligning/updating the todo with existing/resolved issue/discussion numbers/status, de-duplication of todo data, and refactoring, is quick easy work for gen AI, eg prompt with api queried issue and discussion json, and the todo file, and get incremental improvements, small enough for easy PR on a fortnightly, or monthly basis. I'll create an issue and volunteer for that, if one of the vim authors/developers would like me to.)

**Reply by chrisbra** [reply]
If you want to spend time on gen AI, I would rather very much prefer feeding the AI with all the help text so people can ask questions about the documentation directly. Not sure how well this would work, but that is an idea I have for quite a time to explore AI 🤷

**Reply by georgalis** [reply]
> If you want to spend time on gen AI, I would rather very much prefer feeding the AI with all the help text so people can ask questions about the documentation directly. Not sure how well this would work, but that is an idea I have for quite a time to explore AI 🤷

I'm sure that would be straightforward to configure in a "free tier AI wrapper service" ...over 20 years using vim/macros/features, and in the last year, I've learned a lot more about it, just by asking AI with no special prompting. eg I've mapped gc to toggle my config with compatible mode, so I can paste without auto indent, etc, then easily toggle back to my local config.

A url tailored as a web form suitable to support new user vim help queries on a free tier AI, should be easier than deciding which LLM model to use... I'll look into a solution that is easy to link to, as a prerequisite to forum inquiry.

**Reply by nickspoons** [reply]
> toggle my config with compatible mode, so I can paste without auto indent, etc

Vim has `:set paste` for exactly this scenario. Shouldn't an AI have been able to tell you that? This feels like a good example of why we shouldn't rely too much on LLMs for advice just yet...

**Reply by georgalis** [reply]
I used a LLM to craft a support request and submitted it to my LLM vendor (hehe), here is the inquiry as an email template, I encourage anyone interested in this idea to write their LLM vendor of choice, to foster diversity and competition of potential solution providers. (For those unaware, there is a fledgling industry of LLM wrapping services.)

Subject: Exploring LLMs to enhance Vim's user support

As a contributor to the Vim open source project https://www.vim.org/ I'm interested in exploring the use of LLMs to enhance support for our users. Inspired by a recent discussion `https://github.com/vim/vim/discussions/14974#discussioncomment-9755521` I'm considering implementing a web form that allows users to ask Vim-related questions and receive answers from an LLM using our API key. This would show how LLMs can review our documentation and provide helpful responses, encouraging users to create their own accounts to further explore the technology.

1. Do you offer any existing programs or referral bonuses to incentivize open source projects to utilize LLMs for user support and potentially convert their users to subscribers? For instance, a URL that directs users to a system prompt guiding responses relevant to configuring Vim based on the user's request, within the LLM chat interface.

2. If not, could you provide guidance on wrapping the service to achieve this effect? Additionally, what would be the most effective way to prevent potential abuse of our API key?

I believe offering users firsthand experience with LLMs could enhance their learning experience and potentially expand your subscriber base, creating a mutually beneficial relationship.

I would greatly appreciate your insights and recommendations on how to proceed with this idea. I look forward to your response!

Best regards,
(Vim Contributor)

**Reply by georgalis** [reply]
> > toggle my config with compatible mode, so I can paste without auto indent, etc
> 
> Vim has `:set paste` for exactly this scenario. Shouldn't an AI have been able to tell you that? This feels like a good example of why we shouldn't rely too much on LLMs for advice just yet...

I asked LLM for a macro to toggle compatible mode, and got exactly what I asked for. The macro disables hanging indent, word wrap, or nowrap, folding, and whatever else my local config might do, as the file type case may be... so I can copy from vim, reliably, as well.

That said, it is kinda a hack, since I have to maintain a .vimrc-reset file and discard macro redefine errors when I untoggle by resourcing my main config (which resets as well as sets settings). A better solution might be to extend compatible mode with a macro to switch to .vimrc settings, and vice versa, but I'm all set.

**Reply by georgalis** [reply]
> If you want to spend time on gen AI, I would rather very much prefer feeding the AI with all the help text so people can ask questions about the documentation directly. Not sure how well this would work, but that is an idea I have for quite a time to explore AI 🤷

@chrisbra since your comment I've been thinking about how to implement AI support for open source projects. Here is a fearless rephrasing by AI... 

Since your comment, I've been thinking about how to implement AI support for open-source projects. I have also created vendor support inquiries and an issue on the Anthropic Cookbook (anthropics/anthropic-cookbook#60). While I acknowledge the potential concerns about AI leading novice users astray, I believe in the value that a project-oriented system prompt, combined with user questions, could provide. I also emphasize the importance of users knowing how to evaluate solutions, especially when they come from an AI bot.

I've identified three potential approaches worth exploring further:

1. Creating a project API key on a free tier and configuring a web server to submit queries while preventing abuse (e.g., limiting queries to one per IP address every 10 minutes). Although relatively straightforward, this approach isn't ideal and would require root access to the web server.

2. Running a local LLM and allowing unrestricted questions. However, this option might be too slow or costly if using a cloud service.

3. Guiding users to create free-tier accounts with an LLM vendor and providing prompts to help them establish a relationship with the vendor. The exact implementation details for this approach are yet to be determined.

I'm not surprised by the lack of support from vendors, as a solution would require engineering efforts, and the value added is intangible, even for marketing purposes. However, I'm impressed with [RouteLLM](https://github.com/lm-sys/RouteLLM), a framework for identifying the most suitable LLM for a given prompt and routing the query accordingly. RouteLLM's tooling could potentially be refactored for "free tier customer support." Additional links related to this topic can be found in the Anthropic Cookbook issue (anthropics/anthropic-cookbook#60).

**Reply by georgalis** [reply]
> If you want to spend time on gen AI, I would rather very much prefer feeding the AI with all the help text so people can ask questions about the documentation directly. Not sure how well this would work, but that is an idea I have for quite a time to explore AI 🤷

@chrisbra this is now much more viable... a web form can invite noobs to try experimental LLM for support; then prepend a context and role to their question; eg 
`https://claude.ai/new?q="You are a Vim guru well versed in all of the help files and user support forums. Here is your next question, consider the user may be confused by insert verses command mode and all the configuration options, carefully answer considering you do not know their vim familiarity. How do I make vim fold yaml with za?"`

**Reply by chrisbra** [reply]
I will try that one, thanks!

**Comment by georgalis** [comment]
eg, here is a quick AI review of todo for duplicates and consolidation (against my 9.0 patch 1-2122), I would craft a PR for each line item that I could make a reasonable improvement for:

After reviewing the attached todo.txt file, here are a few items that could potentially be grouped together or seem to be duplicates:

Duplicates/Very Similar Items:
- There are two items related to adding a "g)" command to go to the end of a sentence. These could be combined.
- Two items mention supporting ACLs on more systems.
- A couple items mention adding support for underlining and bold modes.
- Two items propose adding a "keepend-contained" argument for syntax regions.

Items that could be grouped:
- Several items related to improving the quickfix list and location list could be grouped under a "Quickfix/Location List Improvements" heading.
- Multiple items related to improving session files and session handling could be grouped.
- Several items discuss improving Vim's printing capabilities on different platforms. These could be grouped under "Printing Improvements".
- A number of items are about improving syntax highlighting performance and capabilities. These could go under a "Syntax Highlighting Enhancements" section.
- Quite a few items relate to improving Vim's command line completion. Grouping them could make sense.
- Some items are about improving the 'path' option handling and could be grouped.

However, most of the items in the file seem distinct enough to stand on their own. The file is also roughly organized by topic already. Some minor grouping or consolidation of very similar items could help, but for the most part the todo list items appear to be clearly separated ideas and features.

**Comment by jiangyinzuo** [comment]
So plain text `todo.txt` (with vim help annotations) has the following advantages:

- independent of any third-party platforms, such as Github, we can freely control the content.
- easy to search/read/edit/organize the content in Vim and other command line tools.
- easy to integrate with gen AI and any other text processing tools.

I believe gen AI is capable of organizing and grouping todo items, but adding and removing todo items (Github issues/PRs) needs to access Github APIs. For removing todo items periodically, we can write a script to check if issues/PRs are closed. For adding todo items periodically,  we might need a rule-based or AI-driven script to filter the active issues/PRs.

[github command line tool](https://cli.github.com/) `gh` might be useful for querying issues/PRs. For example, I tried:

```sh
gh pr list -R vim/vim  --json number --limit 3 -s open --label 'runtime'
```

it outputs
```
[
  {
    "number": 14980
  },
  {
    "number": 14071
  },
  {
    "number": 14057
  }
]
```

**Reply by benknoble** [reply]
This is the kind of thing I would suggest if you want to build tools for comparing GH to todo.txt. The rest seems like noise. (Some form of organized vision/priority/roadmap would be nice, too.)

**Reply by Shane-XB-Qian** [reply]
@benknoble not sure what you were trying to say :sweat_smile:
but todo.txt supposed was just a backlog, why you were thinking it should work like a commercial product?

**Reply by benknoble** [reply]
I certainly didn't suggest it should function like a commercial project. 

I suggest that, if it is valuable for maintainers to compare todo entries with GitHub issues and PRs, they (we) build tools that take advantage of the GH CLI and API. 

I further suggest that maintaining (even in plaintext!) a high-level vision, roadmap, or strategy document might help focus contribution efforts in a unified direction. It's already been pointed out that tracking completion to specific milestones has been useful.

**Reply by jiangyinzuo** [reply]
I think todo.txt should not only include backlog but also high-level vision, roadmap, ...

Actually, todo.txt has priority classification at the top.
```
Priority classification:
9   next point release
8   next release
7   as soon as possible
6   soon
5   should be included
4   nice to have
3   consider including
2   maybe not
1   probably not
-   unclassified
```

However,  recent todo items in todo.txt are all marked as `-` (unclassified).

We need to find a workflow that ensures the todo.txt (or other forms of project management documents) are well maintained without consuming too much effort from the maintainer.

**Reply by Shane-XB-Qian** [reply]
ah... this is a open source project, please donot expect too much, (or not break too much), 
there some history and legend passed, please understand that,
or maybe cannot say todo.txt was not well maintained, i felt ok, but just maybe not your used way.

**Reply by benknoble** [reply]
Just to be clear: I don't think anyone here is saying "the old way sucks." I think we're all saying "What's the path forward? Is there room for improvement? Here are some ideas…"

Rust's contributor flow (just to pick one example) is a testament to having thoughtful conversations and making decisions about to handle social norms for the project.

**Reply by Shane-XB-Qian** [reply]
ok, just to be clear: no one said you cannot improve the room, or ideas, as long as it was meaningful,
you are welcome to contribute more effort and time on this project, i think @vim/admins would thank you.

**Comment by Shane-XB-Qian** [comment]
> * easy to integrate with gen AI

no, it is a "joke" :sweat_smile: 
// as @yegappan had said, todo.txt had its purpose,
// old school had its advantage too, donot need to think web/web/web, or api/api/api etc such things always..
// besides now there github issues/pull-request/discussion etc already, those vs todo.txt was not a conflict thing.

-- 
shane.xb.qian

**Comment by AndrewRadev** [comment]
> Should we adopt better project management methods based on the functionalities of GitHub?

I think it's important to remember that Vim is much, much older than github and has outlasted various different source code hosting providers and version control systems. My personal feeling is that github is way past its peak and is slowly going to be phased out over the next years, but even if you disagree with that, it's useful not to rely on any particular corporation's workflows too much.

Github PRs seem to work well, and maybe it's worth it to use them fully, I'd just be careful to keep a migration path open away from Github/Microsoft.

>  I would rather very much prefer feeding the AI with all the help text so people can ask questions about the documentation directly. Not sure how well this would work, but that is an idea I have for quite a time to explore AI 🤷

Mozilla surprised a number of people by adding an "AI help" button and many developers took to github issues to point out how almost all responses have subtle or not-so-subtle errors. If you're reading the help, presumably it's because you don't know how to do something. If you're then told something incorrect, that's not going to be very helpful in the long run. Here's one github thread if you're curious to read up on the discussion: https://github.com/mdn/yari/issues/9230

My personal experience with teaching Vim included a student that asked me how to define a command with 2 arguments, he said "I tried everything!" and he pasted this ChatGPT-generated code (which I replicated myself):

```
command! -nargs=2 MyCommand call MyFunction(<args>)
```

This includes an obvious error (no such thing as `-nargs=2`) and a subtle error (you should use `<f-args>` for proper escaping). Turns out "I tried everything" did not include `:help command` or reading the slides from my lectures...
